## Data Cleaning - French Baby Names

The goal of this excercise is to use common functions for data cleaning.

### 1. Import libraries

In [1]:
import pandas as pd
import numpy as np

### 2. Read the file nat2018.csv as baby_names and show the head of the data frame

Hint: The separator is a semi-colon " ; "

In [2]:
baby_names = pd.read_csv("nat2018.csv", sep=";")

In [3]:
baby_names.head()

sexe preusuel annais  nombre
0     1        A   1980       3
1     1        A   1998       3
2     1        A   XXXX      21
3     1    AADAM   2009       4
4     1    AADAM   2014       3

The columns are in French, so let's translate them to english to understand the data better.

## 3. Rename column names

The column "sexe" refers to the gender of the baby, "preusuel" is the name given, "annais" is the year the baby was born, and "nombre" is the number of people registered with the same name.
We will rename the columns as following:
- gender
- name
- year
- count

In [4]:
baby_names.columns = ["gender", "name", "year", "count"]

In [5]:
baby_names.head()

gender   name  year  count
0       1      A  1980      3
1       1      A  1998      3
2       1      A  XXXX     21
3       1  AADAM  2009      4
4       1  AADAM  2014      3

Alternatively, rename column names with an explicit method

In [6]:
translations = {"sexe": "gender",
                "preusuel": "name",
                "annais": "year",
                "nombre": "count"}

In [7]:
baby_names.rename(translations, axis=1) # rename along axis 1 (columns)

gender   name  year  count
0            1      A  1980      3
1            1      A  1998      3
2            1      A  XXXX     21
3            1  AADAM  2009      4
4            1  AADAM  2014      3
...        ...    ...   ...    ...
636469       2  ÖZLEM  2013      3
636470       2  ÖZLEM  2014      5
636471       2  ÖZLEM  2015      3
636472       2  ÖZLEM  2016      3
636473       2  ÖZLEM  XXXX     25

[636474 rows x 4 columns]

In [8]:
baby_names = baby_names.rename(columns=translations)

## 4. Check for null values

First, let's look at the length of our data using the shape variable.

In [9]:
baby_names.shape

(636474, 4)

We have 636,474 rows and 4 columns. Now, using the isnull() function with the sum() function, let's look at how many values are null.

In [10]:
baby_names.isnull().sum()

gender    0
name      3
year      0
count     0
dtype: int64

In [11]:
baby_names.head(10)

gender   name  year  count
0       1      A  1980      3
1       1      A  1998      3
2       1      A  XXXX     21
3       1  AADAM  2009      4
4       1  AADAM  2014      3
5       1  AADAM  2016      4
6       1  AADAM  2017      4
7       1  AADAM  2018      3
8       1  AADAM  XXXX      9
9       1  AADEL  1976      5

Retrieve the rows from `baby_names` that have missing names.

---

In [12]:
df = pd.DataFrame([{"name": "nana",
                    "age": 19,
                    "color": "red"},
                   {"name": "bubu",
                    "age": 93,
                    "color": "pink"},
                   {"name": "dede",
                    "age": 29,
                    "color": "blue"},
                   {"name": "kiki",
                    "age": 54,
                    "color": "grey"}])
df

name  age color
0  nana   19   red
1  bubu   93  pink
2  dede   29  blue
3  kiki   54  grey

In [13]:
df["color"] == "pink"

0    False
1     True
2    False
3    False
Name: color, dtype: bool

In [14]:
df.loc[[False, True, False, False]]

name  age color
1  bubu   93  pink

In [15]:
df.loc[df["color"] == "pink"]

name  age color
1  bubu   93  pink

In [16]:
df.loc[df["color"] == "pink", ["name", "age"]]

name  age
1  bubu   93

---

In [17]:
baby_names.isnull()

gender   name   year  count
0        False  False  False  False
1        False  False  False  False
2        False  False  False  False
3        False  False  False  False
4        False  False  False  False
...        ...    ...    ...    ...
636469   False  False  False  False
636470   False  False  False  False
636471   False  False  False  False
636472   False  False  False  False
636473   False  False  False  False

[636474 rows x 4 columns]

In [18]:
baby_names["name"].isnull()

0         False
1         False
2         False
3         False
4         False
          ...  
636469    False
636470    False
636471    False
636472    False
636473    False
Name: name, Length: 636474, dtype: bool

In [19]:
baby_names.loc[baby_names["name"].isnull()]

gender name  year  count
536476       2  NaN  1982      3
536477       2  NaN  2003      3
536478       2  NaN  XXXX     25

In [20]:
baby_names = baby_names.dropna().reset_index(drop=True)

Just three names are null values, so we can remove them from the list. Don't forget to reset the index.

## 5. Verify the data types

In [21]:
baby_names.dtypes

gender     int64
name      object
year      object
count      int64
dtype: object

- Gender and Count are set as integers, which is correct
- Name is an object
- Year should be an integer, but is set as an object because some values are "XXXX"

Let's look at how many rows have "XXXX" set as year.

In [22]:
baby_names.loc[baby_names["year"] == "XXXX"].shape

(35012, 4)

35,012 rows out of the 636,471 have "XXXX" as year. We will remove these rows, since they represent the 9.6% of the data and we will need the year for further analysis.

In [23]:
baby_names = baby_names.loc[baby_names["year"] != "XXXX"].reset_index(drop=True)

Lastly, let's change the year column to integer.

In [24]:
baby_names["year"] = baby_names["year"].astype(int)

In [25]:
baby_names.dtypes

gender     int64
name      object
year       int64
count      int64
dtype: object

We will also change gender to string, 1 being M and 2 being F, using np.where()

In [26]:
baby_names["gender"].value_counts()

2    327476
1    273983
Name: gender, dtype: int64

In [27]:
baby_names["gender"]

0         1
1         1
2         1
3         1
4         1
         ..
601454    2
601455    2
601456    2
601457    2
601458    2
Name: gender, Length: 601459, dtype: int64

In [28]:
# np.where(condition, then, else)
np.where(baby_names["gender"] == 1, "M", "F")

array(['M', 'M', 'M', ..., 'F', 'F', 'F'], dtype='<U1')

In [40]:
baby_names = baby_names.assign(gender=np.where(baby_names["gender"] == 1, "M", "F"))

In [30]:
baby_names

gender   name  year  count
0           M      A  1980      3
1           M      A  1998      3
2           M  AADAM  2009      4
3           M  AADAM  2014      3
4           M  AADAM  2016      4
...       ...    ...   ...    ...
601454      F  ÖZLEM  2012      6
601455      F  ÖZLEM  2013      3
601456      F  ÖZLEM  2014      5
601457      F  ÖZLEM  2015      3
601458      F  ÖZLEM  2016      3

[601459 rows x 4 columns]

## 6. How many unique names does our list have?

Using nunique() for the "name" column, look at how many unique names out data has.

In [31]:
#this code groups by name and finds the unique values of "count"
baby_names.groupby("name")["count"].unique()

name
A                                              [3]
AADAM                                       [4, 3]
AADEL                                    [5, 3, 4]
AADIL                [4, 6, 8, 9, 11, 12, 3, 7, 5]
AAHIL                                    [5, 4, 6]
                           ...                    
ÖMER     [3, 4, 5, 13, 23, 29, 35, 39, 56, 59, 64]
ÖZGE                                     [4, 3, 5]
ÖZGÜR                                          [3]
ÖZKAN                                          [3]
ÖZLEM                           [3, 7, 4, 8, 6, 5]
Name: count, Length: 31708, dtype: object

In [32]:
baby_names["name"].nunique()

31708

We have 31,708 unique names. Now using value_counts() for the "name" column, let's look at how many babies have the same name.

In [33]:
baby_names["name"].value_counts()

ALIX              238
_PRENOMS_RARES    238
CAMILLE           238
ANGE              238
DOMINIQUE         235
                 ... 
MINETTE             1
JELICA              1
ROSEMAINE           1
SHEURLEY            1
NISAR               1
Name: name, Length: 31708, dtype: int64

There are 238 values with the name "\_PRENOMS_RARES", which translates to rare names, so we will remove this names from the data.

In [34]:
#renamed in case we want to access old data
baby_names_clean_names = baby_names.loc[baby_names["name"]!="_PRENOMS_RARES"].reset_index(drop=True).copy()

In [35]:
baby_names_clean_names.head(5)

gender   name  year  count
0      M      A  1980      3
1      M      A  1998      3
2      M  AADAM  2009      4
3      M  AADAM  2014      3
4      M  AADAM  2016      4

In [36]:
#gives a visual confirmation that rows were removed
baby_names_clean_names["name"].value_counts()

ANGE         238
CAMILLE      238
ALIX         238
DOMINIQUE    235
CLAUDE       232
            ... 
SASHKA         1
FANELIA        1
ROLAN          1
HETAN          1
NISAR          1
Name: name, Length: 31707, dtype: int64

## 7. Get the top 10 most common baby names
Using the value counts for the "name" column, sort the values in descending order and save the top 10 in a variable called top_names.

In [37]:
top_names = baby_names_clean_names["name"].value_counts().head(10)

In [38]:
top_names

ANGE         238
CAMILLE      238
ALIX         238
DOMINIQUE    235
CLAUDE       232
HYACINTHE    228
MAXIME       226
MARIE        211
GABY         206
CYRILLE      206
Name: name, dtype: int64

#### i. Let's get a separate list of top names for males and for females

Group by gender using count as an aggregate function to see how many values we have of each.

In [63]:
(baby_names_clean_names.groupby(by=["gender", "name"])
                         .agg({"count":"sum"})
                         .sort_values(by=["gender","count"], ascending=False))

count
gender name              
M      JEAN       1916584
       PIERRE      890973
       MICHEL      820274
       ANDRÉ       711933
       PHILIPPE    538373
...                   ...
F      ZOHOUR           3
       ZOPHIE           3
       ZOUBEIDA         3
       ZUBIDA           3
       ÉLIZABETH        3

[33680 rows x 1 columns]

In [ ]:
baby_names_clean_names.loc[baby_names_clean_names["gender"]=="M"]

Now, create a data frame where gender is M named male_names and another for female_names. Compare the length of each data frame with the previous data frame in order to be sure that we have the same values.

In [65]:
male_names = baby_names_clean_names.loc[baby_names_clean_names["gender"]=="M"].copy()

In [66]:
female_names = baby_names_clean_names.loc[baby_names_clean_names["gender"]=="F"].copy()

In [68]:
male_names.head()

gender   name  year  count
0      M      A  1980      3
1      M      A  1998      3
2      M  AADAM  2009      4
3      M  AADAM  2014      3
4      M  AADAM  2016      4

In [69]:
female_names.head()

gender    name  year  count
273864      F   AALIA  2014      3
273865      F   AALIA  2015      3
273866      F   AALIA  2017      7
273867      F   AALIA  2018     10
273868      F  AALIYA  2002      7

#### ii. Get the top 10 names for males and for females

In [74]:
#this tells us how many times the name appears in the dataframe, but does not aggregate the count of the names
#we need to use a groupby
male_names["name"].value_counts().sort_values(ascending=False)

CHRISTIAN    119
ARNAUD       119
GEORGES      119
HUGUES       119
ALFRED       119
            ... 
TOUFIRK        1
KYMANY         1
LIEVEN         1
IKE            1
RATANA         1
Name: name, Length: 15304, dtype: int64

In [81]:
(male_names.groupby(by="name")
             .agg({"count":"sum"})
             .sort_values(by="count", ascending=False).head(10))

count
name             
JEAN      1916584
PIERRE     890973
MICHEL     820274
ANDRÉ      711933
PHILIPPE   538373
LOUIS      517608
RENÉ       516337
ALAIN      506865
JACQUES    482669
BERNARD    469295

## Bonus

### 1. Create a data frame containing babies born from 2010 onwards

### 2. Get the top 10 most common names from the data frame created.

### 3.  From the `baby_names` dataframe, calculate the registered babies for each year

### 4. From the `baby_names` dataframe, calculate the registered babies for each gender and each year

### 5. For each year, what are the top 5 names?

### 6. For each year and gender, what are the top 5 names?

### 7. What's the ratio of female babies for each year?

### 8. Identify all unique unisex names